File: LinkedList/Easy/203. Remove Linked List Elements/203. Remove Linked List Elements.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * Removes all nodes with the given value from a linked list.
 *
 * @param head Pointer to the head of the linked list.
 * @param val The value to remove from the list.
 * @return The head of the modified linked list.
 *
 * Time Complexity: O(n) - where n is the number of nodes in the linked list. 
 * We traverse the list once to remove nodes.
 *
 * Space Complexity: O(1) - No extra space is used apart from the pointers.
 */
struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode *temp;
    
    // Remove nodes from the beginning of the list that match the value
    while (head != NULL && head->val == val) {
        head = head->next;
    }
    
    // If the list is empty after removal, return NULL
    if (head == NULL) {
        return NULL;
    }
    
    // Traverse the list and remove nodes with the given value
    temp = head;
    while (temp->next) {
        if (temp->next->val == val) {
            temp->next = temp->next->next;
        } else {
            temp = temp->next;
        }
    }
    
    return head;
}

File: LinkedList/Easy/203. Remove Linked List Elements/203. Remove Linked List Elements.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def removeElements(self, head, val):
        """
        Removes all nodes with the given value from a linked list.

        :param head: ListNode, the head of the linked list.
        :param val: Integer, the value to be removed from the list.
        :return: ListNode, the head of the modified linked list.
        """
        # Remove nodes from the beginning of the list that match the value
        while head and head.val == val:
            head = head.next
        
        # If the list is empty after removal, return None
        if head is None:
            return head
        
        # If the list has only one node
        if head.next is None:
            if head.val != val:
                return head
            return None
        
        # Traverse the list and remove nodes with the given value
        temp = head
        while temp.next:
            if temp.next.val == val:
                temp.next = temp.next.next
            else:
                temp = temp.next
        
        return head

# Example usage:
# head = ListNode(1, ListNode(2, ListNode(6, ListNode(3, ListNode(4, ListNode(5, ListNode(6)))))))
# solution = Solution()
# new_head = solution.removeElements(head, 6)
# The new_head should not contain any nodes with value 6
